File: deprecated/truth-studies/truth-matching-studies/mc16_matching/studies/top/topkinematics.py
from AnalysisG.Templates import SelectionTemplate

class TopKinematics(SelectionTemplate):

    def __init__(self):
        SelectionTemplate.__init__(self)
        self.res_top_kinematics = {"pt" : [], "eta" : [], "phi" : [], "energy" : []}
        self.spec_top_kinematics = {"pt" : [], "eta" : [], "phi" : [], "energy" : []}
        self.deltaR = {"SS" : [], "RS" : [], "RR" : []}
        self.mass_combi = {"SS" : [], "RS" : [], "RR" : []}

    def Selection(self, event):
        tops = event.Tops
        if len([t for t in tops if t.FromRes == 1]) != 2: return False
        if len([t for t in tops if t.FromRes == 0]) != 2: return False
        return True

    def Strategy(self, event):
        tops = event.Tops

        rtops = [t for t in tops if t.FromRes == 1]
        self.res_top_kinematics["pt"]     += [r.pt/1000 for r in rtops]
        self.res_top_kinematics["eta"]    += [r.eta for r in rtops]
        self.res_top_kinematics["phi"]    += [r.phi for r in rtops]
        self.res_top_kinematics["energy"] += [r.e/1000 for r in rtops]

        stops = [t for t in tops if t.FromRes == 0]
        self.spec_top_kinematics["pt"]     += [r.pt/1000 for r in stops]
        self.spec_top_kinematics["eta"]    += [r.eta for r in stops]
        self.spec_top_kinematics["phi"]    += [r.phi for r in stops]
        self.spec_top_kinematics["energy"] += [r.e/1000 for r in stops]

        self.deltaR["RS"] += [r.DeltaR(s) for r in rtops for s in stops]
        self.deltaR["SS"] += [stops[0].DeltaR(stops[1])]
        self.deltaR["RR"] += [rtops[0].DeltaR(rtops[1])]

        self.mass_combi["RS"] += [(r + s).Mass/1000 for r in rtops for s in stops]
        self.mass_combi["SS"] += [(stops[0] + stops[1]).Mass/1000]
        self.mass_combi["RR"] += [(rtops[0] + rtops[1]).Mass/1000]

File: src/AnalysisG/selections/mc16/topkinematics/topkinematics.cxx
#include "topkinematics.h"

topkinematics::topkinematics(){this -> name = "top-kinematics";}
topkinematics::~topkinematics(){}

selection_template* topkinematics::clone(){
    return (selection_template*)new topkinematics();
}

void topkinematics::merge(selection_template* sl){
    topkinematics* slt = (topkinematics*)sl; 

    merge_data(&this -> res_top_kinematics , &slt -> res_top_kinematics); 
    merge_data(&this -> spec_top_kinematics, &slt -> spec_top_kinematics); 
    merge_data(&this -> mass_combi         , &slt -> mass_combi); 
    merge_data(&this -> deltaR             , &slt -> deltaR); 
}

bool topkinematics::selection(event_template* ev){
    bsm_4tops* evn = (bsm_4tops*)ev; 
    std::vector<particle_template*>* tops = &evn -> Tops; 
    int res = 0; 
    for (size_t x(0); x < tops -> size(); ++x){res += ((top*)tops -> at(x)) -> from_res;} 
    return res == 2; 
}

bool topkinematics::strategy(event_template* ev){
    auto kins = [](std::map<std::string, std::vector<float>>* data, std::vector<top*>* td){
        for (size_t x(0); x < td -> size(); ++x){
            (*data)["pt"].push_back(td -> at(x) -> pt / 1000); 
            (*data)["eta"].push_back(td -> at(x) -> eta); 
            (*data)["phi"].push_back(td -> at(x) -> phi); 
            (*data)["energy"].push_back(td -> at(x) -> e / 1000); 
        }
    };

    bsm_4tops* evn = (bsm_4tops*)ev; 
    std::vector<top*> r_tops = {};
    std::vector<top*> s_tops = {}; 
    for (size_t x(0); x < evn -> Tops.size(); ++x){
        top* t = (top*)evn -> Tops[x]; 
        if (t -> from_res){r_tops.push_back(t);}
        else {s_tops.push_back(t);}
    }

    kins(&this -> res_top_kinematics , &r_tops); 
    kins(&this -> spec_top_kinematics, &s_tops); 

    std::vector<float>* data = nullptr; 

    data = &this -> deltaR["RS"]; 
    for (size_t x(0); x < r_tops.size(); ++x){
        for (size_t y(0); y < s_tops.size(); ++y){
            data -> push_back(r_tops[x] -> DeltaR(s_tops[y])); 
        }
    }

    data = &this -> deltaR["RR"]; 
    for (size_t x(0); x < r_tops.size(); ++x){
        for (size_t y(x); y < r_tops.size(); ++y){
            if (x == y){continue;}
            data -> push_back(r_tops[x] -> DeltaR(r_tops[y])); 
        }
    }

    data = &this -> deltaR["SS"]; 
    for (size_t x(0); x < s_tops.size(); ++x){
        for (size_t y(x); y < s_tops.size(); ++y){
            if (x == y){continue;}
            data -> push_back(s_tops[x] -> DeltaR(s_tops[y])); 
        }
    }
 
    data = &this -> mass_combi["RS"]; 
    for (size_t x(0); x < r_tops.size(); ++x){
        for (size_t y(0); y < s_tops.size(); ++y){
            top r = (*(r_tops[x]) + *s_tops[y]); 
            data -> push_back(r.mass / 1000); 
        }
    }

    data = &this -> mass_combi["RR"]; 
    for (size_t x(0); x < r_tops.size(); ++x){
        for (size_t y(x); y < r_tops.size(); ++y){
            if (x == y){continue;}
            top r = (*(r_tops[x]) + *r_tops[y]); 
            data -> push_back(r.mass / 1000); 
        }
    }

    data = &this -> mass_combi["SS"]; 
    for (size_t x(0); x < s_tops.size(); ++x){
        for (size_t y(x); y < s_tops.size(); ++y){
            if (x == y){continue;}
            top r = (*(s_tops[x]) + *s_tops[y]); 
            data -> push_back(r.mass / 1000); 
         }
    }

    return true; 
}
